### PR TITLE
Enabling higher prefill batch size in Granite4.0

### PR DESCRIPTION
**Pull request by ksmusz**
# Enable Higher Prefill Batch Size for Granite 4.0 MambaMixer2

## Summary

Removes hard-coded BS=1 prefill limitations across the Granite MambaMixer2 conv1d, SSM, and model runner code paths, enabling multi-sequence prefill batches on Gaudi HPUs.

## Changes (5 files)

- **`granite_causal_conv1d.py`** — Remove BS=1 assertion; add batched conv1d path with per-sequence state gather/scatter.
- **`hpu_mamba_mixer2.py`** — Fix output reshape for dynamic batch shape; add padded `cu_seqlens` and `chunks_per_sequence` computation for multi-batch SSM.
- **`hpu_model_runner.py`** — Fix `last_chunk_indices` to use per-sequence offsets: `(i+1)*nphysical_chunks - 1`.
- **`ssd_combined.py`** — Thread `chunks_per_sequence` through SSM forward and varlen wrapper.
- **`pytorch_implementation.py`** — Add `chunks_per_sequence` to `new_ssd_state_passing()` and `new_chunk_scan()` for independent per-sequence processing in batched layout.

## Backward Compatibility

- BS=1 path is unchanged and preserved.
- No new configuration flags required — the model runner naturally batches when scheduler provides multiple prefill requests.

### CODE REVIEW

File: vllm_gaudi/v1/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -2420,7 +2420,11 @@ def _form_prefill_batch(self, contents):
             assert chunk_size > 0
             nphysical_chunks = target_seq // chunk_size
             assert nphysical_chunks > 0, (f"target_seq={target_seq} must be >= chunk_size={chunk_size}")
-            last_chunk_indices = [nphysical_chunks - 1 for _ in range(len(contents.req_ids))]
+            # For multi-batch prefill (BS>1), the SSM processes
+            # BS * target_seq tokens.  Each sequence occupies
+            # nphysical_chunks chunks in the flattened layout, so
+            # sequence i's last chunk is at (i+1)*nphysical_chunks - 1.
+            last_chunk_indices = [(i + 1) * nphysical_chunks - 1 for i in range(len(contents.req_ids))]
```
[the remark this answers is not in the document]
Comment (ksmusz) [reply vllm_gaudi/v1/worker/hpu_model_runner.py]: Applied

File: vllm_gaudi/ops/granite_causal_conv1d.py [hunk 2]
```diff
@@ -55,26 +55,48 @@ def granite_causal_conv1d_fn(
...
+        x_batch = x_work.view(dim, padded_batch, target_seq).permute(1, 0, 2)
+
+        # Load per-sequence init states: [BS, state_len, dim]
+        if has_initial_state is not None:
+            raw_states = conv_states[cache_indices, -state_len:, :]  # [BS, state_len, dim]
+            mask = has_initial_state.view(-1, 1, 1).to(dtype=raw_states.dtype)
+            init_states = raw_states * mask  # zero out where no initial state
+        else:
+            init_states = torch.zeros(padded_batch, state_len, dim, device=x_work.device, dtype=work_dtype)
+        init_states = init_states.transpose(-1, -2)  # [BS, dim, state_len]
```
[the remark this answers is not in the document]
Comment (ksmusz) [reply vllm_gaudi/ops/granite_causal_conv1d.py:92]: Applied

File: vllm_gaudi/ops/granite_causal_conv1d.py [hunk 3]
```diff
@@ -55,26 +55,48 @@ def granite_causal_conv1d_fn(
...
+
+        # Gather new states per-sequence at actual_len positions
+        actual_lens = (qsl[1:padded_batch + 1] - qsl[:padded_batch]).clamp(min=0)
+        col_offsets = torch.arange(state_len, device=x_work.device, dtype=torch.int64)
+        col_indices = (actual_lens.unsqueeze(-1).to(torch.int64) + col_offsets.unsqueeze(0))  # [BS, state_len]
+        col_indices = col_indices.unsqueeze(1).expand(-1, dim, -1)
+        new_states = torch.gather(seq_input, 2, col_indices)  # [BS, dim, state_len]
+
+        # Store new conv states (skip padding entries with actual_len==0)
+        conv_states[cache_indices, -state_len:, :] = new_states.transpose(-1, -2)
```
[the remark this answers is not in the document]
Comment (ksmusz) [reply vllm_gaudi/ops/granite_causal_conv1d.py]: Applied partially.
What was done: before the scatter into `conv_states`, the per-sequence
`new_states_T` rows are zeroed out for inactive slots (`actual_lens == 0`)
via `torch.where(active_mask, new_states_T, 0)`. With duplicate PAD
cache indices this makes every duplicate write store the same (zero)
value, so the result is deterministic regardless of which duplicate
"wins".

What was deliberately not done: we did not filter
`cache_indices`/`new_states_T` down to only the active rows with
boolean indexing. That would introduce a data-dependent (dynamic) shape
into the write, which is costly on HPU (breaks graph reuse / bucket
compatibility). The zero-constant-value approach keeps the write shape
static and still eliminates the nondeterminism the reviewer flagged.
